ext/node/resolution.rs

```diff
@@ -706,41 +716,63 @@ fn file_exists(path: &Path) -> bool {
 pub fn legacy_main_resolve(
   package_json: &PackageJson,
   referrer_kind: NodeModuleKind,
-) -> Result<PathBuf, AnyError> {
-  let maybe_main = package_json.main(referrer_kind);
+  conditions: &[&str],
+) -> Result<Option<PathBuf>, AnyError> {
+  let is_types = conditions == TYPES_CONDITIONS;
+  let maybe_main = if is_types {
+    package_json.types.as_ref()
+  } else {
+    package_json.main(referrer_kind)
+  };
   let mut guess;
 
   if let Some(main) = maybe_main {
     guess = package_json.path.parent().unwrap().join(main).clean();
     if file_exists(&guess) {
-      return Ok(guess);
+      return Ok(Some(guess));
     }
 
     let mut found = false;
     // todo(dsherret): investigate exactly how node handles this
-    let endings = match referrer_kind {
-      NodeModuleKind::Cjs => vec![
-        ".js",
-        ".cjs",
-        ".json",
-        ".node",
-        "/index.js",
-        "/index.cjs",
-        "/index.json",
-        "/index.node",
-      ],
-      NodeModuleKind::Esm => vec![
-        ".js",
-        ".mjs",
-        ".json",
-        ".node",
-        "/index.js",
-        "/index.mjs",
-        ".cjs",
-        "/index.cjs",
-        "/index.json",
-        "/index.node",
-      ],
+    let endings = if is_types {
+      match referrer_kind {
+        NodeModuleKind::Cjs => {
+          vec![".d.ts", ".d.cts", "/index.d.ts", "/index.d.cts"]
+        }
+        NodeModuleKind::Esm => vec![
+          ".d.ts",
+          ".d.mts",
+          "/index.d.ts",
+          "/index.d.mts",
+          ".d.cts",
+          "/index.d.cts",
+        ],
+      }
+    } else {
+      match referrer_kind {
+        NodeModuleKind::Cjs => vec![
+          ".js",
+          ".cjs",
+          ".json",
+          ".node",
+          "/index.js",
+          "/index.cjs",
+          "/index.json",
+          "/index.node",
+        ],
+        NodeModuleKind::Esm => vec![
+          ".js",
+          ".mjs",
+          ".json",
+          ".node",
+          "/index.js",
+          "/index.mjs",
+          ".cjs",
+          "/index.cjs",
+          "/index.json",
+          "/index.node",
```

Comment:
These seems to be wrong

Reply:
What's wrong? I just copied from the previous code.

Reply:
I don't understand why ESM is trying `.cjs`, `.json` and `.node` extensions, thought these are only done when using `require()`

Reply:
This may be an esm referrer importing cjs. I just tried it out in node and looks like it never probes for a cjs extension or even mjs, which is nice and require doesn't even probe for index.cjs. I'll remove these for now.